src/OrleansGoogleUtils/Providers/Streams/GooglePubSubAdapterReceiver.cs

```diff
@@ -0,0 +1,150 @@
+            }
+        }
+
+        public async Task<IList<IBatchContainer>> GetQueueMessagesAsync(int maxCount)
+        {
+            try
+            {
+                var pubSubRef = _pubSub; // store direct ref, in case we are somehow asked to shutdown while we are receiving.    
+                if (pubSubRef == null) return new List<IBatchContainer>();
+
+                int count = maxCount < 0 ? 10 : maxCount;
+
+                var task = pubSubRef.GetMessages(count);
+                _outstandingTask = task;
+                IEnumerable<ReceivedMessage> messages = await task;
+
+                List<IBatchContainer> pubSubMessages = new List<IBatchContainer>();
+                foreach (var message in messages)
+                {
+                    IBatchContainer container = _dataAdapter.FromPullResponseMessage(message.Message, _lastReadMessage++);
+                    pubSubMessages.Add(container);
+                    _pending.Add(new PendingDelivery(container.SequenceToken, message));
+                }
+
+                return pubSubMessages;
+            }
+            finally
+            {
+                _outstandingTask = null;
+            }
+        }
+
+        public async Task MessagesDeliveredAsync(IList<IBatchContainer> messages)
+        {
+            try
+            {
+                var pubSubRef = _pubSub; // store direct ref, in case we are somehow asked to shutdown while we are receiving.
+                if (messages.Count == 0 || pubSubRef == null) return;
+                // get sequence tokens of delivered messages
+                List<StreamSequenceToken> deliveredTokens = messages.Select(message => message.SequenceToken).ToList();
+                // find oldest delivered message
+                StreamSequenceToken oldest = deliveredTokens.Max();
+                // finalize all pending messages at or befor the oldest
+                List<PendingDelivery> finalizedDeliveries = _pending
+                    .Where(pendingDelivery => !pendingDelivery.Token.Newer(oldest))
+                    .ToList();
+                if (finalizedDeliveries.Count == 0) return;
+                // remove all finalized deliveries from pending, regardless of if it was delivered or not.
+                _pending.RemoveRange(0, finalizedDeliveries.Count);
+                // get the queue messages for all finalized deliveries that were delivered.
+                List<ReceivedMessage> deliveredMessages = finalizedDeliveries
+                    .Where(finalized => deliveredTokens.Contains(finalized.Token))
+                    .Select(finalized => finalized.Message)
+                    .ToList();
+                if (deliveredMessages.Count == 0) return;
+                // delete all delivered queue messages from the queue.  Anything finalized but not delivered will show back up later
+                _outstandingTask = Task.WhenAll(deliveredMessages.Select(m => pubSubRef.AcknowledgeMessages(new[] { m })));
+
```

Comment:
both google's native api and your `GooglePubSubDataManager` has a batch operation for AcknowledgeMessages. Why do you use it like a single operation? This would potentially bring performance penalty when you try to acknowledge a large amount of messages